### ISSUE TITLE
Move registering `callback_fns` with Learner from `.fit()` to `__post_init__()`.

### ISSUE THREAD

**Issue by pechyonkin** [fastai/fastai#2235]
**TLDR**: when trying to run `.get_preds()` right after creating a learner, `callback_fns` are not working because they are registered with the learner in the `.fit()` function. I propose to move registering `callback_fns` to initialization of learner so that callbacks can work even without training first.

I am trying to use a pretrained model and analyze activations for some data; in other words, I am not using training, only predicting from an existing model.

I create the data, then the learner, and a test hook callback that prints when a batch is done:

```python
class TestHook(HookCallback):
    def hook(self, m, i, o):
        pass
    def on_batch_end(self, train, **kwargs):
        print("Batch Done!")
        
learner = cnn_learner(data, models.resnet18, callback_fns=TestHook)
```

When I run `learner.fit_one_cycle(1)`, the callback is invoked after each batch and `"Batch Done!" is printed correctly.

However, when I run `learner.get_preds(data.train_ds)`, nothing happens. After digging into the source code, I found the reason: callbacks are registered with the learner in the [`fit()`](https://github.com/fastai/fastai/blob/bd430ef40e7759057aec3e607e7f7834fe26a110/fastai/basic_train.py#L198) function. This means that after creating the learner, callbacks are still not registered and therefore will not work when running `.get_preds()` immediately after creating the learner.

I was able to solve this the following way, and callback works correctly, but this looks too hacky:

```python
class TestHook(HookCallback):
    def hook(self, m, i, o):
        pass
    def on_batch_end(self, train, **kwargs):
        print("Batch Done!")     
        
learner = cnn_learner(data, models.resnet18)
test_hook_callback = TestHook(learner)
learner.callbacks += [test_hook_callback]

learner.get_preds(data.train_ds)
```

Please let me know if this is worth pursuing. If yes, I can try to change the code and create a pull request.

Some more context is in this [post](https://forums.fast.ai/t/callbacks-for-prediction-step/50667) of fastai forums.

**Comment by sgugger** [comment]
We can certainly try to move this to post_init and see if it breaks anything. It would make more sense to have them initialized there.

**Comment by pechyonkin** [comment]
I will work on it and will submit a PR when ready.

**Comment by pechyonkin** [comment]
I did some research and just moving callback registering creates conflicts and failing tests. I have described some of them [on the forums](https://forums.fast.ai/t/moving-registering-callback-fns-with-learner-from-fit-to-post-init/50946).
